catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergS3Signer.java

```diff
@@ -0,0 +1,194 @@
+import org.projectnessie.catalog.service.api.CatalogService;
+import org.projectnessie.catalog.service.api.EncryptionService;
+import org.projectnessie.catalog.service.api.SnapshotReqParams;
+import org.projectnessie.catalog.service.api.SnapshotResponse;
+import org.projectnessie.error.NessieContentNotFoundException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.IcebergView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Immutable
+abstract class IcebergS3Signer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IcebergS3Signer.class);
+
+  abstract IcebergS3SignRequest request();
+
+  abstract ParsedReference ref();
+
+  abstract ContentKey key();
+
+  abstract String encryptedLocation();
+
+  abstract CatalogService catalogService();
+
+  abstract EncryptionService encryptionService();
+
+  abstract RequestSigner signer();
+
+  abstract S3BucketOptions s3options();
+
+  @Value.Lazy
+  String requestedS3Uri() {
+    return S3Utils.asS3Location(request().uri());
+  }
+
+  @Value.Lazy
+  String decryptedLocation() {
+    return encryptionService().decrypt(encryptedLocation());
+  }
+
+  @Value.Lazy
+  boolean write() {
+    return request().method().equalsIgnoreCase("PUT")
+        || request().method().equalsIgnoreCase("POST")
+        || request().method().equalsIgnoreCase("DELETE");
+  }
+
+  Uni<IcebergS3SignResponse> verifyAndSign() {
+    return fetchSnapshot()
+        .call(this::checkForbiddenLocations)
+        .onItem()
+        .transformToMulti(this::collectAllowedLocations)
+        .filter(requestedS3Uri()::startsWith)
+        .toUni()
+        .onItem()
+        .ifNull()
+        .failWith(this::unauthorized)
+        .replaceWith(request().uri())
+        .map(this::sign);
+  }
+
+  private Uni<SnapshotResponse> fetchSnapshot() {
+    try {
+      CompletionStage<SnapshotResponse> stage =
+          catalogService()
+              .retrieveSnapshot(
+                  SnapshotReqParams.forSnapshotHttpReq(ref(), "iceberg", null), key(), null);
+      // consider an import failure as a non-existing content:
+      // signing will be authorized for the future location only.
+      return Uni.createFrom().completionStage(stage).onFailure().recoverWithNull();
+    } catch (NessieContentNotFoundException ignored) {
+      return Uni.createFrom().nullItem();
+    } catch (NessieNotFoundException e) {
+      return Uni.createFrom().failure(e);
+    }
+  }
+
+  private Uni<?> checkForbiddenLocations(SnapshotResponse snapshotResponse) {
+    if (snapshotResponse != null && write()) {
+      Content content = snapshotResponse.content();
+      String metadataLocation = null;
+      if (content instanceof IcebergTable) {
+        metadataLocation = ((IcebergTable) content).getMetadataLocation();
```

Comment:
Actually, this can be simplified by checking for the object-name-pattern for metadata-files.

Reply:
Is it carved in stone? Isn't it better to trust whatever was saved in the database?

Reply:
The pattern for the metadata-JSONs is pretty much defined, yes.
Checking the pattern prevents overwriting metadata.